### PR TITLE
[Metricbeat] VSphere module dashboards

### PR DESCRIPTION
**Pull request by philippkahr**
Hi,

I created some dashboards for the VSphere module.

https://github.com/elastic/beats/issues/14101

### CODE REVIEW

**Comment by philippkahr** [comment]
The Travis build is failing with 

```text
Get the licenses available from ['./vendor', './metricbeat/vendor', './metricbeat/module/kvm/vendor', './metricbeat/module/postgresql/vendor', './metricbeat/module/vsphere/vendor', './metricbeat/module/mysql/vendor']
```

but I do not know how to fix this.

File: metricbeat/module/vsphere/_meta/kibana/7/dashboard/metricbeat-vsphere-host.json [hunk 1]
```diff
@@ -0,0 +1,1286 @@
...
+                "show": true,
+                "style": {},
+                "title": {
+                  "text": "Free"
+                },
+                "type": "value"
+              }
+            ]
+          },
+          "title": "Free vs Used CPU on Cluster stacked [metricbeat VSphere] ECS",
```
Comment (ChrsMark) [inline metricbeat/module/vsphere/_meta/kibana/7/dashboard/metricbeat-vsphere-host.json]:
```suggestion
          "title": "Free vs Used CPU on Cluster stacked [Metricbeat VSphere] ECS",
```

File: metricbeat/module/vsphere/_meta/kibana/7/dashboard/metricbeat-vsphere-host.json [hunk 2]
```diff
@@ -0,0 +1,1286 @@
...
+        "kibanaSavedObjectMeta": {
+          "searchSourceJSON": {
+            "filter": [],
+            "query": {
+              "language": "kuery",
+              "query": ""
+            }
+          }
+        },
+        "title": "top datastore used [Metricbeat VSphere] ECS",
```
Comment (ChrsMark) [inline metricbeat/module/vsphere/_meta/kibana/7/dashboard/metricbeat-vsphere-host.json]:
```suggestion
        "title": "Top datastore used [Metricbeat VSphere] ECS",
```

File: metricbeat/module/vsphere/_meta/kibana/7/dashboard/metricbeat-vsphere-vm.json [hunk 3]
```diff
@@ -0,0 +1,850 @@
...
+          "searchSourceJSON": {
+            "filter": [],
+            "query": {
+              "language": "kuery",
+              "query": ""
+            }
+          }
+        },
+        "savedSearchRefName": "search_0",
+        "title": "how many VMs run on a ESXI host [Metricbeat VSphere] ECS",
```
Comment (ChrsMark) [inline metricbeat/module/vsphere/_meta/kibana/7/dashboard/metricbeat-vsphere-vm.json]:
```suggestion
        "title": "VMs per ESXI host [Metricbeat VSphere] ECS",
```

File: metricbeat/module/vsphere/_meta/kibana/7/dashboard/metricbeat-vsphere-vm.json [hunk 4]
```diff
@@ -0,0 +1,850 @@
...
+          "searchSourceJSON": {
+            "filter": [],
+            "query": {
+              "language": "kuery",
+              "query": ""
+            }
+          }
+        },
+        "savedSearchRefName": "search_0",
+        "title": "OS Pie [Metricbeat VSphere] ECS",
```
Comment (ChrsMark) [inline metricbeat/module/vsphere/_meta/kibana/7/dashboard/metricbeat-vsphere-vm.json]:
```suggestion
        "title": "OS Distribution [Metricbeat VSphere] ECS",
```

**Comment by philippkahr** [comment]
Hi @ChrsMark 

thanks for reviewing my changes. Sorry for the "typos", maybe we should add a naming scheme and example to the contributions guide?

> 1. Could you add an updated screenshot of the dashboards in this PR so as people can comment on the visualizations and propose improvements to make them more meaningful?

Sure! I will do that later tonight.

> 2. I wonder if it is ok to have 2 different dashboards for this module.
Hmm, that depends really. I think the first dashboard is really for the cluster and hosts. The other one is a drill down into the stats of the virtual machine. Combining those would not make much sense.

**Comment by ChrsMark** [comment]
> Hi @ChrsMark
> 
> thanks for reviewing my changes. Sorry for the "typos", maybe we should add a naming scheme and example to the contributions guide?
> 

Sure! We are currently in discussions about improving the Developer Guides in general. Feel free to open an issue mentioning that proposal.

**Comment by ChrsMark** [comment]
Thank you for taking the time to open the issue @philippkahr!

We can go on with this PR regardless, so feel free to add screenshot(s) of the dashboards in new comments here so as to verify them and maybe suggest improvements. 🙂

**Comment by Jarzamendia** [comment]
I'm very excited to see this in action! My team are very happy with the Infrastructure dashboard. If we can also see the VSphere metrics to, this stack will add so much value to our environment!

**Comment by philippkahr** [comment]
Hi,

sorry, it took so long.

I had to add a lot of white boxes and some example text since I am using real data. I hope that this is still ok.

## Metricbeat VSphere  Hosts Overview ECS
Metricbeat VSphere  Hosts Overview ECS

## Metricbeat VSphere  VMs overview ECS
Metricbeat VSphere  VMs overview ECS

The heatmap itself is used to show where a VM has been on an ESXi host in the timeframe. This might show you if a VM is likely to be moved around a lot. The other dashboards are self-explanatory, to me at least :).

**Comment by Jarzamendia** [comment]
A very good view would be a VM distributed by the hosts. Similar to what you see on the Infrastructure tab. This would allow us to quickly identify hosts that could soon become overloaded if we know which VMs will soon have access spikes.

File: metricbeat/module/vsphere/_meta/kibana/7/dashboard/metricbeat-vsphere-vm.json [hunk 5]
```diff
@@ -0,0 +1,850 @@
...
+          "searchSourceJSON": {
+            "filter": [],
+            "query": {
+              "language": "kuery",
+              "query": ""
+            }
+          }
+        },
+        "savedSearchRefName": "search_0",
+        "title": "VM running on what ESXi host [Metricbeat VSphere] ECS",
```
Comment (ChrsMark) [inline metricbeat/module/vsphere/_meta/kibana/7/dashboard/metricbeat-vsphere-vm.json]:
```suggestion
        "title": "VMs' placements on ESXi Hosts[Metricbeat VSphere] ECS",
```

File: metricbeat/module/vsphere/_meta/kibana/7/dashboard/metricbeat-vsphere-vm.json [hunk 6]
```diff
@@ -0,0 +1,850 @@
...
+                "scale": {
+                  "defaultYExtents": false,
+                  "type": "linear"
+                },
+                "show": false,
+                "type": "value"
+              }
+            ]
+          },
+          "title": "VM running on what ESXi host [Metricbeat VSphere] ECS",
```
Comment (ChrsMark) [inline metricbeat/module/vsphere/_meta/kibana/7/dashboard/metricbeat-vsphere-vm.json]:
```suggestion
          "title": "VM placements on ESXi Hosts[Metricbeat VSphere] ECS",
```

File: metricbeat/module/vsphere/_meta/kibana/7/dashboard/metricbeat-vsphere-vm.json [hunk 7]
```diff
@@ -0,0 +1,850 @@
...
+          "searchSourceJSON": {
+            "filter": [],
+            "query": {
+              "language": "kuery",
+              "query": ""
+            }
+          }
+        },
+        "savedSearchRefName": "search_0",
+        "title": "networks from virtualmachine [Metricbeat VSphere] ECS",
```
Comment (ChrsMark) [inline metricbeat/module/vsphere/_meta/kibana/7/dashboard/metricbeat-vsphere-vm.json]: Not so sure what this chart shows but maybe we could change `virtualmachine` to `VM`?
```suggestion
        "title": "Networks from virtualmachine [Metricbeat VSphere] ECS",
```

File: metricbeat/module/vsphere/_meta/kibana/7/dashboard/metricbeat-vsphere-vm.json [hunk 8]
```diff
@@ -0,0 +1,850 @@
...
+            "labels": {
+              "last_level": true,
+              "show": false,
+              "truncate": 100,
+              "values": true
+            },
+            "legendPosition": "right",
+            "type": "pie"
+          },
+          "title": "networks from virtualmachine [Metricbeat VSphere] ECS",
```
Comment (ChrsMark) [inline metricbeat/module/vsphere/_meta/kibana/7/dashboard/metricbeat-vsphere-vm.json]:
```suggestion
          "title": "Networks from virtualmachine [Metricbeat VSphere] ECS",
```

File: metricbeat/module/vsphere/_meta/kibana/7/dashboard/metricbeat-vsphere-vm.json [hunk 9]
```diff
@@ -0,0 +1,850 @@
...
+            "showMetricsAtAllLevels": false,
+            "showPartialRows": false,
+            "showTotal": false,
+            "sort": {
+              "columnIndex": null,
+              "direction": null
+            },
+            "totalFunc": "sum"
+          },
+          "title": "virtualmachine used CPU total ram network [Metricbeat VSphere] ECS",
```
Comment (ChrsMark) [inline metricbeat/module/vsphere/_meta/kibana/7/dashboard/metricbeat-vsphere-vm.json]:
```suggestion
          "title": "CPU/RAM/Network per VM [Metricbeat VSphere] ECS",
```

File: metricbeat/module/vsphere/_meta/kibana/7/dashboard/metricbeat-vsphere-host.json [hunk 10]
```diff
@@ -0,0 +1,1286 @@
...
+        "kibanaSavedObjectMeta": {
+          "searchSourceJSON": {
+            "filter": [],
+            "query": {
+              "language": "kuery",
+              "query": ""
+            }
+          }
+        },
+        "title": "Top datastore used [Metricbeat VSphere] ECS",
```
Comment (ChrsMark) [inline metricbeat/module/vsphere/_meta/kibana/7/dashboard/metricbeat-vsphere-host.json]:
```suggestion
        "title": "Top Datastore Used [Metricbeat VSphere] ECS",
```

File: metricbeat/module/vsphere/_meta/kibana/7/dashboard/metricbeat-vsphere-host.json [hunk 11]
```diff
@@ -0,0 +1,1286 @@
...
+                "terms_field": "vsphere.datastore.name",
+                "terms_order_by": "5aec7160-f047-11e9-8ffe-6377df51c23d"
+              }
+            ],
+            "show_grid": 1,
+            "show_legend": 1,
+            "time_field": "",
+            "type": "gauge"
+          },
+          "title": "top datastore used [Metricbeat VSphere] ECS",
```
Comment (ChrsMark) [inline metricbeat/module/vsphere/_meta/kibana/7/dashboard/metricbeat-vsphere-host.json]:
```suggestion
          "title": "Top Datastore Used [Metricbeat VSphere] ECS",
```

File: metricbeat/module/vsphere/_meta/kibana/7/dashboard/metricbeat-vsphere-host.json [hunk 12]
```diff
@@ -0,0 +1,1286 @@
...
+          "searchSourceJSON": {
+            "filter": [],
+            "query": {
+              "language": "kuery",
+              "query": ""
+            }
+          }
+        },
+        "savedSearchRefName": "search_0",
+        "title": "Datastore types [Metricbeat VSphere] ECS",
```
Comment (ChrsMark) [inline metricbeat/module/vsphere/_meta/kibana/7/dashboard/metricbeat-vsphere-host.json]:
```suggestion
        "title": "Datastore Types [Metricbeat VSphere] ECS",
```

**Comment by ChrsMark** [comment]
Hi @philippkahr ! Thank you for adding the screenshots. :)

In general it looks good to me! I just added some minor suggestions so as to keep the namings consistent.

**Comment by ChrsMark** [comment]
> A very good view would be a VM distributed by the hosts. Similar to what you see on the Infrastructure tab. This would allow us to quickly identify hosts that could soon become overloaded if we know which VMs will soon have access spikes.

I think that the Heatmap @philippkahr adds can give this information, since it actually shows the placements of the VMs on each Host.

**Comment by kaiyan-sheng** [comment]
Sorry for the late review. This dashboard looks great! Could you please also document it in `beats/metricbeat/module/vsphere/_meta/docs.asciidoc`? Something like: 
```
[float]
=== Dashboard

The vsphere module comes with a predefined dashboard. For example:

image::./images/metricbeat_vsphere_dashboard.png[]
```
This will also require to put a screenshot of this dashboard `metricbeat_vsphere_dashboard.png` into `beats/metricbeat/docs/images`. Then it will show up in documentation for vsphere module 😄

**Comment by philippkahr** [comment]
@ChrsMark thanks for the suggestions

@kaiyan-sheng I've added the information as requested :)

**Comment by ChrsMark** [comment]
@philippkahr great! you will need again `make update` since you changed the docs. Also if didn't already, please rebase on top of the latest master.

**Comment by philippkahr** [comment]
Hi @ChrsMark 

can you please double-check if I did the rebase correctly? I am doing this for the first time. 

```
inside the vspheredashboard branch from my forked repo
$ git remote -v

origin  git@github.com:philippkahr/beats.git (fetch)
origin  git@github.com:philippkahr/beats.git (push)
upstream        https://github.com/elastic/beats.git (fetch)
upstream        https://github.com/elastic/beats.git (push) 

$ git rebase upstream/master

First, rewinding head to replay your work on top of it...
Applying: https://github.com/elastic/beats/issues/14101 renamed everything according to xyz [metricbeat] ECS, imported and exported the dashboards, created the module.yml
Applying: https://github.com/elastic/beats/pull/14135 added dashboard
Applying: https://github.com/elastic/beats/pull/14135 trying to fix the error from travis ci
Applying: https://github.com/elastic/beats/pull/14135 fixing naming pattern missmatch
Applying: https://github.com/elastic/beats/pull/14135: fixed: empty description, fixed: expected title with format 'Some title [Metricbeat Module]'
Applying: https://github.com/elastic/beats/pull/14135: travis told me to run `Error: some files are not up-to-date. Run 'mage fmt update' then review and commit the changes. Modified: [metricbeat/docs/modules_list.asciidoc]`
Applying: https://github.com/elastic/beats/pull/14135 Remove submodule beats from vendor
Applying: Update metricbeat/module/vsphere/_meta/kibana/7/dashboard/metricbeat-vsphere-vm.json
Applying: Update metricbeat/module/vsphere/_meta/kibana/7/dashboard/metricbeat-vsphere-vm.json
Applying: Update metricbeat/module/vsphere/_meta/kibana/7/dashboard/metricbeat-vsphere-host.json
Applying: Update metricbeat/module/vsphere/_meta/kibana/7/dashboard/metricbeat-vsphere-host.json
Applying: Update metricbeat/module/vsphere/_meta/kibana/7/dashboard/metricbeat-vsphere-vm.json
Applying: Update metricbeat/module/vsphere/_meta/kibana/7/dashboard/metricbeat-vsphere-vm.json
Applying: Update metricbeat/module/vsphere/_meta/kibana/7/dashboard/metricbeat-vsphere-vm.json
Applying: Update metricbeat/module/vsphere/_meta/kibana/7/dashboard/metricbeat-vsphere-host.json
Applying: Update metricbeat/module/vsphere/_meta/kibana/7/dashboard/metricbeat-vsphere-host.json
Applying: Update metricbeat/module/vsphere/_meta/kibana/7/dashboard/metricbeat-vsphere-host.json
Applying: Update metricbeat/module/vsphere/_meta/kibana/7/dashboard/metricbeat-vsphere-vm.json
Applying: Update metricbeat/module/vsphere/_meta/kibana/7/dashboard/metricbeat-vsphere-vm.json
Applying: https://github.com/elastic/beats/pull/14135 added the vsphere dashboards to the docs
Applying: make update

$ git push -f origin vspheredashboards 

Enumerating objects: 211, done.
Counting objects: 100% (211/211), done.
Delta compression using up to 8 threads
Compressing objects: 100% (161/161), done.
Writing objects: 100% (197/197), 1.78 MiB | 2.33 MiB/s, done.
Total 197 (delta 138), reused 10 (delta 3)
remote: Resolving deltas: 100% (138/138), completed with 13 local objects.
To github.com:philippkahr/beats.git
 + 13de30c44...41c36d5a3 vspheredashboards -> vspheredashboards (forced update)

```

**Comment by ChrsMark** [comment]
> Hi @ChrsMark
> 
> can you please double-check if I did the rebase correctly? I am doing this for the first time.
> 
> ```
> inside the vspheredashboard branch from my forked repo
> $ git remote -v
> 
> origin  git@github.com:philippkahr/beats.git (fetch)
> origin  git@github.com:philippkahr/beats.git (push)
> upstream        https://github.com/elastic/beats.git (fetch)
> upstream        https://github.com/elastic/beats.git (push) 
> 
> $ git rebase upstream/master
> 
> First, rewinding head to replay your work on top of it...
> Applying: https://github.com/elastic/beats/issues/14101 renamed everything according to xyz [metricbeat] ECS, imported and exported the dashboards, created the module.yml
> Applying: https://github.com/elastic/beats/pull/14135 added dashboard
> Applying: https://github.com/elastic/beats/pull/14135 trying to fix the error from travis ci
> Applying: https://github.com/elastic/beats/pull/14135 fixing naming pattern missmatch
> Applying: https://github.com/elastic/beats/pull/14135: fixed: empty description, fixed: expected title with format 'Some title [Metricbeat Module]'
> Applying: https://github.com/elastic/beats/pull/14135: travis told me to run `Error: some files are not up-to-date. Run 'mage fmt update' then review and commit the changes. Modified: [metricbeat/docs/modules_list.asciidoc]`
> Applying: https://github.com/elastic/beats/pull/14135 Remove submodule beats from vendor
> Applying: Update metricbeat/module/vsphere/_meta/kibana/7/dashboard/metricbeat-vsphere-vm.json
> Applying: Update metricbeat/module/vsphere/_meta/kibana/7/dashboard/metricbeat-vsphere-vm.json
> Applying: Update metricbeat/module/vsphere/_meta/kibana/7/dashboard/metricbeat-vsphere-host.json
> Applying: Update metricbeat/module/vsphere/_meta/kibana/7/dashboard/metricbeat-vsphere-host.json
> Applying: Update metricbeat/module/vsphere/_meta/kibana/7/dashboard/metricbeat-vsphere-vm.json
> Applying: Update metricbeat/module/vsphere/_meta/kibana/7/dashboard/metricbeat-vsphere-vm.json
> Applying: Update metricbeat/module/vsphere/_meta/kibana/7/dashboard/metricbeat-vsphere-vm.json
> Applying: Update metricbeat/module/vsphere/_meta/kibana/7/dashboard/metricbeat-vsphere-host.json
> Applying: Update metricbeat/module/vsphere/_meta/kibana/7/dashboard/metricbeat-vsphere-host.json
> Applying: Update metricbeat/module/vsphere/_meta/kibana/7/dashboard/metricbeat-vsphere-host.json
> Applying: Update metricbeat/module/vsphere/_meta/kibana/7/dashboard/metricbeat-vsphere-vm.json
> Applying: Update metricbeat/module/vsphere/_meta/kibana/7/dashboard/metricbeat-vsphere-vm.json
> Applying: https://github.com/elastic/beats/pull/14135 added the vsphere dashboards to the docs
> Applying: make update
> 
> $ git push -f origin vspheredashboards 
> 
> Enumerating objects: 211, done.
> Counting objects: 100% (211/211), done.
> Delta compression using up to 8 threads
> Compressing objects: 100% (161/161), done.
> Writing objects: 100% (197/197), 1.78 MiB | 2.33 MiB/s, done.
> Total 197 (delta 138), reused 10 (delta 3)
> remote: Resolving deltas: 100% (138/138), completed with 13 local objects.
> To github.com:philippkahr/beats.git
>  + 13de30c44...41c36d5a3 vspheredashboards -> vspheredashboards (forced update)
> ```

Yeah it looks ok, this is the way to go with rebasing. Thank you!

**Comment by ChrsMark** [comment]
ok to test

File: metricbeat/module/vsphere/_meta/kibana/7/dashboard/metricbeat-vsphere-vm.json [hunk 13]
```diff
@@ -0,0 +1,850 @@
...
+          "searchSourceJSON": {
+            "filter": [],
+            "query": {
+              "language": "kuery",
+              "query": ""
+            }
+          }
+        },
+        "savedSearchRefName": "search_0",
+        "title": "virtualmachine used CPU total ram network [Metricbeat VSphere] ECS",
```
Comment (ChrsMark) [inline metricbeat/module/vsphere/_meta/kibana/7/dashboard/metricbeat-vsphere-vm.json]:
```suggestion
        "title": "CPU/RAM/Network per VM [Metricbeat VSphere] ECS",
```

**Comment by ChrsMark** [comment]
Loaded the dashboards successfully. Found one left-over on one of the tittles.

After that I think we are ok to merge.

**Comment by ChrsMark** [comment]
ok to test

**Comment by ChrsMark** [comment]
Failing tests, seem to be irrelevant. Merging.

Thank you so much for contributing this @philippkahr 🙂 !

**Pull request merged**